gpflow/models/model.py

```diff
@@ -283,11 +239,31 @@ def predict_y(
         return self.likelihood.predict_mean_and_var(f_mean, f_var)
 
     def predict_log_density(
-        self, data: RegressionData, full_cov: bool = False, full_output_cov: bool = False
+        self, data: RegressionData, full_cov: bool = False, full_output_cov: bool = False,
     ):
         """
         Compute the log density of the data at the new data points.
         """
         X, Y = data
         f_mean, f_var = self.predict_f(X, full_cov=full_cov, full_output_cov=full_output_cov)
         return self.likelihood.predict_density(f_mean, f_var, Y)
+
+
+class InternalDataGPModel(GPModel, InternalDataTrainingInterface):
+    """
+    Base class for models that encapsulate their data and store it as
+    self.data; training_loss does not take any arguments.
+    """
+
+    def training_loss(self):
+        return self.maximum_a_posteriori_objective()
```

Comment:
and flip the sign here as well as in the other one below, which I still think is a bit messy :/

Reply:
I think I'd prefer to keep the `_training_loss()` and `_training_loss(data)`...

Reply:
i think this is much better. i think have a method which is _training_loss is much messier.